src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1184 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+
+use mz_repr::adt::datetime::DateTimeUnits;
+use mz_repr::{ColumnType, Datum, RelationType, Row, RowArena, ScalarType};
+
+use crate::{
+    BinaryFunc, EvalError, MapFilterProject, MfpPlan, MirScalarExpr, UnaryFunc,
+    UnmaterializableFunc, VariadicFunc,
+};
+
+/// Tracks whether a function is monotonic, either increasing or decreasing.
+/// This property is useful for static analysis because a monotonic function maps ranges to ranges:
+/// ie. if `a` is between `b` and `c` inclusive, then `f(a)` is between `f(b)` and `f(c)` inclusive.
+/// (However, if either `f(b)` and `f(c)` error or return null, we don't assume anything about `f(a)`.
+///
+/// This should likely be moved to a function annotation in the future.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum Monotonic {
+    /// Monotonic in this argument.
```

Comment:
nit: `Monotonically increasing or decreasing` (I think otherwise folks would only assume the other when consulting the docs).

Reply:
A header comment that explains exactly what we mean (e.g., pointwise, non-strict) with some examples would be very helpful to our future selves.

Reply:
Good call! I added a few comments, here and on some related functions. (And I promise to carry these comments forward if and when these become annotations!)